Tortorella add api to save user progress data

Jtortorella opened the pull request.

Jtortorella committed:
Had to delete some of the work.
With the updated content services it just was too different and would be difficult to merge.
This is the updated neutral api hook.

Jtortorella committed:
Sorted leaderboard in the leaderboard service.
May need to follow up for current leaderboard implementation.
Made new users start lesson at 1, will need to follow up once UI changes are implemented.

char-lock commented on backend/src/endpoints/users/users.controller.ts:
```diff
@@ -114,7 +114,7 @@ export default class UsersController {
       validation_method: validation_method,
       validation_value: validation_value,
-      progress_lesson: 0,
+      progress_lesson: 1,
```
Either both need to be zero, or both need to be 1. Either is okay, but we want to be consistent.

Jtortorella replied:
Updated progress_lesson to 0.

char-lock commented on backend/src/endpoints/users/users.controller.ts:
```diff
@@ -215,5 +215,39 @@ export default class UsersController {
+      contentProgress,
+    } = req.body;
+    logger.debug(`request recieved: get user with username ${username}`);
+    if (contentProgress < 10)
```
We cannot always be certain that a lesson will contain exactly 10 pieces of content.

Jtortorella replied:
Passed questionList.length to the api body.

char-lock commented on backend/src/endpoints/users/users.controller.ts:
```diff
@@ -215,5 +215,39 @@ export default class UsersController {
+      response = UsersData.updateContentProgressByUsername(username);
+    }
+    else {
+      response = UsersData.updateLessonProgressByUsername(username);
```
Are we just assuming that the value is changing by one? Consider cases when an educator sets multiple lessons as skipped or content is skipped.

char-lock commented on backend/src/endpoints/users/users.data.ts:
```diff
@@ -121,4 +121,34 @@ export default class UsersData {
+      },
+      data: {
+        progress_content: {
+          increment: 1,
```
See comment re: assuming a single increment.

char-lock commented on backend/src/endpoints/users/users.data.ts:
```diff
@@ -121,4 +121,34 @@ export default class UsersData {
+      data: {
+        progress_content: 0,
+        progress_lesson: {
+          increment: 1,
```
See comment re: assuming a single increment.

char-lock commented on backend/src/endpoints/users/users.routes.ts:
```diff
@@ -19,4 +19,7 @@ router.get("/validation/:value", UsersController.getUsersByValidationValue);
 router.post("/", UsersController.createUser);
 
+// Updates a user's score.
+router.post("/username/:username/updateprogress", UsersController.updateUserProgress);
```
Never, ever set an endpoint to an action like this. It should be a noun -- /username/:username/progress, with GET returning the values and POST/PUT/PATCH/whatever writing the values, and potentially DELETE clearing them to baseline.

Jtortorella replied:
changed function name to postUserProgressByUsername.

changed function in UsersData to postContentProgressByUsername, postLessonProgressByUsername (resp.)

Jtortorella replied:
Hey sorry I forgot to push changes.

char-lock commented on projekt-monsterz-api:
I've removed this from the last two pulls; remove this submodule.

Jtortorella replied:
Removed submodule.

char-lock commented on ui/src/app/pages/dashboard/dashboard.component.ts:
```diff
@@ -97,8 +97,11 @@ export class DashboardComponent implements OnInit {
     return (this.globalLeaderboardSelected)
-      ? this.leaderboardService.GetGlobalLeaderboard()
-      : this.leaderboardService.GetClassLeaderboard();
+      ? this.leaderboardService.GetGlobalLeaderboard().sort((a, b) => {
```
Could be more concise by just sorting the ternary result:

```
return (
  this.globalLeaderboardSelected 
    ? this.leaderboardService.GetGlobalLeaderboard() 
    : this.leaderboardService.GetClassLeaderboard()
  ).sort((a, b) => b.score - a.score);

Jtortorella replied:
Updated logic. To ternary expression.

char-lock commented on ui/src/app/services/leaderboard.service.ts:
```diff
@@ -58,16 +58,27 @@ export class LeaderboardService {
+          (score) => {
+            this.leaderboardGlobal.push(score);
+            if (this.leaderboardGlobal.length === users.length) { this.updatingGlobal = false; }
+          });
```
Interesting. Consider adding this as a single line filter to the callback (i.e.,

```
if (score.username === this._user.username) score.username = "YOU";

Jtortorella replied:
Changed logic. Tested and it works fine.

Jtortorella replied:
see getUserScores in api.service. Logic is moved to callback.

Jtortorella committed:
Updated progress_lesson to 0.
Passed questionList.length to the api body. Will be used instead of assuming contentProgress is always 10.
changed function name to postUserProgressByUsername in users.routes.ts
changed function in UsersData to postContentProgressByUsername, postLessonProgressByUsername (resp.)
removed api and projekt-monsterz-api submodules.
Updated logic to ternary expression in dashboard.component.ts.
see getUserScores in api.service. Logic is moved to callback to change current user's name to you.
Need to discuss logic in users.data.ts, I might have a better solution.

Jtortorella committed:
Updated progress_lesson to 0.
Passed questionList.length to the api body. Will be used instead of assuming contentProgress is always 10.
changed function name to postUserProgressByUsername.
changed function in UsersData to postContentProgressByUsername, postLessonProgressByUsername (resp.)
Removed submodule api and projekt monsterz api.
see getUserScores in api.service. Logic is moved to callback.

Jtortorella committed:
Updated progress_lesson to 0.
Passed questionList.length to the api body. Will be used instead of assuming contentProgress is always 10.
changed function name to postUserProgressByUsername.
changed function in UsersData to postContentProgressByUsername, postLessonProgressByUsername (resp.)
Removed submodule api and projekt monsterz api.
see getUserScores in api.service. Logic is moved to callback.

Jtortorella committed:
Updated Comment

Jtortorella committed:
Updated backend to include class endpoints to:
1. Get all students from a class
2. Remove / Add a student from a class.
3. Create / Destroy a class.
Verified through postman every endpoint works.
Updated schema to add class code property on the user.
Updated schema to include a class table that contained the ID, Instructor and date created.
Will need to revisit the ID as it was not randomly generated as well as I would hope.
Might need to check into some verification, right now I don't think the remove user from class is working as restrictive as I want it to.

Jtortorella committed:
Okay checked that remove student works.
Messed more with the automatic generation of class id.
Update types for class.
Updated schema.

Jtortorella committed:
Updated endpoints to be more in line with demo goals.
Rolled back original idea to have a boolean in content scheme and update progress.

Jtortorella committed:
Updated endpoints and models.
Removed boolean from content schema and removed associated changes.

Jtortorella committed:
Verified endpoints worked.

Jtortorella committed:
Checked all routes and verified connectivity and functionality.
